feat(router): set stricter type for Route.title

markostanimirovic opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

The type of `Route.title` is `string | unknown`.

Issue Number: #44936


## What is the new behavior?

The type of `Route.title` is `string | Type<Resolve<string>>`.


## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

AFAIK, the route title feature is not released yet. So there will be no breaking changes.

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->

alfaproject commented on packages/router/test/page_title_strategy_spec.ts:
```diff
@@ -101,10 +101,10 @@ describe('title strategy', () => {
                {
                  path: '',
-                 // `null` prevents inheriting from parent if it doesn't have its own value.
+                 // `undefined` prevents inheriting from parent if it doesn't have its own value.
```
This doesn't sound right. So if you want to inherit from parent, do you have to do it manually in all children?

markostanimirovic replied:
AFAIK there is no need to explicitly set `title` to `undefined` if you want to inherit the title from the parent.

So, with the following route config:

```ts
const routes: Routes = [
  {
    path: 'foo',
    component: FooComponent,
    title: 'FOO',
    children: [
      { path: 'bar', component: BarComponent }
    ]
  },
]
```

the title for route `foo/bar` will be `FOO`.

atscott commented on packages/router/src/config.ts:
```diff
@@ -9,6 +9,7 @@
 import {NgModuleFactory, NgModuleRef, Type} from '@angular/core';
 import {Observable} from 'rxjs';
 
+import {Resolve} from './interfaces';
```
Totally reasonable. The specific locations of the types don't really matter since they're all just exported from the index.

I'd recommend doing this as a separate commit (same PR) to make things easier to review.

markostanimirovic committed:
refactor(router): move types from config and interfaces to models

This commit moves all types from `config.ts` and `interfaces.ts`
to `models.ts` to fix circular dependency.

markostanimirovic committed:
feat(router): set stricter type for Route.title

This commit changes the type of `Route.title` property from
`string | unknown` to `string | Type<Resolve<string>>`.

Fixes #44936

atscott reviewed (approved):
LGTM 👍  Thanks for the improvement!

reviewed-for: fw-router
reviewed-for: public-api

public API review notes: 
As I mentioned in the issue report, I do prefer keeping this as `Type<Resolve<string>>` rather than creating a type alias that would be another click-through in the documentation just to arrive at the same `Resolve<string>` type. I don't think the type alias would be useful in this situation and would actually make things _more_ obscure.

jessicajaniuk reviewed (approved):
LGTM :cookie:

reviewed-for: public-api

atscott commented:
caretaker note: this will require an update to one of the patches in g3: [cl/426173402](http://cl/426173402)

AndrewKushnir reviewed (approved):
Reviewed-for: public-api

dylhunn commented:
This PR was merged into the repository by commit 791bd3142432ed7caf14a68a6e434b27004e634b.
